lib/Infer/ExhaustiveSynthesis.cpp

```diff
@@ -0,0 +1,567 @@
+// experiment with synthesizing at reduced bitwidth, then expanding the result
+// aggressively avoid calling into the solver
+
+// TODO move this to Inst
+const bool isCmp(Inst::Kind K) {
+  return K == Inst::Eq || K == Inst::Ne || K == Inst::Ult ||
+    K == Inst::Slt || K == Inst::Ule || K == Inst::Sle;
+}
+
+// TODO move this to Inst
+const bool isShift(Inst::Kind K) {
+  return K == Inst::Shl || K == Inst::AShr || K == Inst::LShr;
+}
+
+void hasConstantHelper(Inst *I, std::set<Inst *> &Visited,
+                       std::vector<Inst *> &ConstList) {
+  // FIXME this only works for one constant and keying by name is bad
+  if (I->K == Inst::Var && (I->Name.find("reserved_") != std::string::npos)) {
+    // FIXME use a less stupid sentinel
+    ConstList.push_back(I);
+  } else {
+    if (Visited.insert(I).second)
+      for (auto Op : I->Ops)
+        hasConstantHelper(Op, Visited, ConstList);
+  }
+}
+
+// TODO do this a more efficient way
+// TODO do this a better way, checking for constants by name is dumb
+void hasConstant(Inst *I, std::vector<Inst *> &ConstList) {
+  std::set<Inst *> Visited;
+  hasConstantHelper(I, Visited, ConstList);
+}
+
+// TODO small vector by reference?
+std::vector<Inst *> matchWidth(Inst *I, unsigned NewW, InstContext &IC) {
+  int OldW = isCmp(I->K) ? 1 : I->Width;
+  if (OldW > NewW)
+    return { IC.getInst(Inst::Trunc, NewW, { I }) };
+  if (OldW < NewW)
+    return {
+            IC.getInst(Inst::SExt, NewW, { I }),
+            IC.getInst(Inst::ZExt, NewW, { I }),
+    };
+  return { I };
+}
+
+void addGuess(Inst *RHS, int MaxCost, std::vector<Inst *> &Guesses,
+              int &TooExpensive) {
+  if (souper::cost(RHS) < MaxCost)
+    Guesses.push_back(RHS);
+  else
+    TooExpensive++;
+}
+
+void findVars(Inst *Root, std::vector<Inst *> &Vars) {
+  // breadth-first search
+  std::set<Inst *> Visited;
+  std::queue<Inst *> Q;
+  Q.push(Root);
+  while (!Q.empty()) {
+    Inst *I = Q.front();
+    Q.pop();
+    if (!Visited.insert(I).second)
+      continue;
+    if (I->K == Inst::Var)
+      Vars.push_back(I);
+    for (auto Op : I->Ops)
+      Q.push(Op);
+  }
+}
+
+void getGuesses(std::vector<Inst *>& Guesses,
+                std::vector<Inst* >& Vars,
+                std::vector<Inst *>& Inputs,
+                int Width, int LHSCost,
+                InstContext &IC) {
+
+  int TooExpensive = 0;
+  // start with the nops -- but not a constant since that is
+  // legitimately faster to synthesize using the special-purpose code
+  for (auto I : Inputs) {
+    auto v = matchWidth(I, Width, IC);
+    for (auto N : v)
+      addGuess(N, LHSCost, Guesses, TooExpensive);
+  }
+
+  // TODO enforce permitted widths
+  // TODO try both the source and dest width, if they're different
+  std::vector<Inst::Kind> Unary = {
+    Inst::CtPop, Inst::BSwap, Inst::Cttz, Inst::Ctlz
+  };
+  if (Width > 1) {
+    for (auto K : Unary) {
+      for (auto I : Inputs) {
+        auto v1 = matchWidth(I, Width, IC);
```

Comment:
please capitalize local variables like v1 also.
or maybe rename this one to be more informative?
I know you got this code from me, but it was bad when I wrote it, sorry